### PR TITLE
[close #96] Remove all absolute paths from cache

### LINKED ISSUE #96: Sprockets::FileOutsidePaths error seems to be caused by fix to #59
Not a 100% sure, but what seems to be happening is even though cache key generation now uses relative paths, some or all depedency references are absolute.

so if you are sharing the asset cache folder among multiple release, sprockets ends up finding the right cache entry but then tries to load dependencies from an older release absolute path (where the cache was originally constructed) and fails with Sprockets::FileOutsidePaths

### PR DESCRIPTION
**Pull request by schneems**
[close #96] Remove all absolute paths from cache

The patch in #92 was incomplete, it converted all cache keys to use relative paths, but didn't fully remove all absolute paths from cache contents. The test case was accidentally passing since we didn't check to make sure any of the paths from cache were different from the original ones stored. This patch eliminates these absolute paths stored in the cache:

- [x] dependency paths
- [x] filename
- [x] asset uris
- [x] "included" paths (no idea what these are)
- [x] load paths

### URITar

This patch works by introducing a utility class URITar which can "compress" or "expand" a given uri. A "compressed" uri can either be one that is relative to the root, or an absolute path if it is outside of the root. An expanded uri will always be an absolute path.

A uri that is relative to the root will be compressed with no beginning slash

    file://relative/to/root/file.js

A uri that is outside of the root will be compressed with a beginning slash

    file:///absolute/path/to/file.js

Even though I'm using "uri" here the URITar class can also operate on file paths without a valid URI scheme name. Like:

    relative/to/root/file.js

and

    /absolute/path/to/file.js

The UnloadedAsset class was moved to it's own file and refactored to use the new URITar class.

## Fix

Before putting anything in the cache, we will "compress" all uris  and paths so that no absolute paths are in the cache (unless they're not relative to the root which would indicate they're somewhere global e.g. from a gem or shared directory). 

Upon loading an asset in memory, we "expand" all uris since sprockets relies on absolute paths for just about everything. 

Almost all the business logic is limited to the loader, so the rest of sprockets has no clue if relative or absolute paths were used to build the asset.

We are also compressing the "environment-paths" so that dependencies in different paths will differ. I think this is needed, but the tests don't fail when it's taken out.

ATP

### CODE REVIEW

File: lib/sprockets.rb [hunk 1]
```diff
@@ -147,7 +147,9 @@ module Sprockets
     env.paths
   end
   register_dependency_resolver 'file-digest' do |env, str|
-    env.file_digest(env.parse_file_digest_uri(str))
+    env.file_digest( # lib/sprockets/base.rb:47
```
Comment (rafaelfranca) [inline lib/sprockets.rb]: Why this comment? Do we have more than one sprockets env?

File: lib/sprockets/uri_utils.rb [hunk 2]
```diff
@@ -137,15 +138,24 @@ def parse_file_digest_uri(uri)
...
+    #   # => "file-digest:///tmp/js/application.js"
     #
     # path - String file path
     #
     # Returns String URI.
     def build_file_digest_uri(path)
-      join_file_uri('file-digest'.freeze, nil, path, nil)
+      relative_path =  UnloadedAsset.new(path, self).relative_path
+      if relative_path == path
+        # path is absolute, use forward slash
```
Comment (rafaelfranca) [inline lib/sprockets/uri_utils.rb]: You mean relative right?

File: lib/sprockets/uri_utils.rb [hunk 3]
```diff
@@ -137,15 +138,24 @@ def parse_file_digest_uri(uri)
     # Examples
     #
     #   build("/tmp/js/application.js")
-    #   # => "file-digest:/tmp/js/application.js"
+    #   # => "file-digest:///tmp/js/application.js"
```
Comment (rafaelfranca) [inline lib/sprockets/uri_utils.rb]: Why not use `file-digest:/tmp/js/application.js` for absolute paths and `file-digest:tmp/js/application.js` for relative?

File: lib/sprockets.rb (hunk 1 above)
Comment (schneems) [inline lib/sprockets.rb]: We have more than one `file_digest` :frowning:. This PR is really messy, needs a major-major cleanup

File: lib/sprockets/uri_utils.rb (hunk 3 above)
Comment (rafaelfranca) [inline lib/sprockets/uri_utils.rb]: Ah, we were already using this pattern you only fixed the method :+1:

Comment (schneems) [inline lib/sprockets/uri_utils.rb]: :boom: I like that better. There's a few checks to make sure these are valid "uris' but we can probably get rid of some of those checks.

File: lib/sprockets/uri_utils.rb (hunk 2 above)
Comment (schneems) [inline lib/sprockets/uri_utils.rb]: UnloadedAsset#relative_path may return an absolute path if the `uri `passed into it is outside of the `environment.root`. So when the original path matches the "relative path" it means that the "relative path" returned is in fact absolute. It's not ideal naming, i'm working on it.

**Comment by elia** [comment]
Included paths can be inlined multiple times 

> ### The include Directive
> include path works like require, but inserts the contents of the specified source file even if it has already been included or required.

_https://github.com/rails/sprockets/blob/2.x/README.md_

Don't know if it has been removed or deprecated in recent versions tho

**Comment by schneems** [comment]
PR is cleaned up and updated. Pretty sure this is good to go, however 3.x is failing. Want to wait until we get everything passing again on travis to be sure. Once this is merged i'll cut 3.3.2.

**Comment by schneems** [comment]
Build is green: https://travis-ci.org/rails/sprockets/builds/76162275

File: lib/sprockets/loader.rb [hunk 4]
```diff
@@ -303,16 +175,46 @@ def load_from_unloaded(unloaded)
...
         # Save the asset in the cache under the new URI
-        cache.set(stored_asset.asset_key, asset, true)
+        cached_asset             = asset.dup
+        cached_asset[:uri]       = compress_from_root(asset[:uri])
+        cached_asset[:filename]  = compress_from_root(asset[:filename])
+        cached_asset[:load_path] = compress_from_root(asset[:load_path])
+
+        if cached_asset[:metadata]
+          # Deep dup to avoid modifying `asset`
+          cached_asset[:metadata]                = cached_asset[:metadata].dup
```
Comment (rafaelfranca) [inline lib/sprockets/loader.rb]: fix the space here

File: lib/sprockets/loader.rb [hunk 5]
```diff
@@ -303,16 +175,46 @@ def load_from_unloaded(unloaded)
...
+          if cached_asset[:metadata][:included]
+            cached_asset[:metadata][:included] = cached_asset[:metadata][:included].dup
```
Comment (rafaelfranca) [inline lib/sprockets/loader.rb]: Why this dup

File: lib/sprockets/unloaded_asset.rb [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    #
+    # Returns a String.
+    def filename
+      unless @filename
+        load_file_params
+      end
+      @filename
+    end
+
+
```
Comment (rafaelfranca) [inline lib/sprockets/unloaded_asset.rb]: :scissors:

File: lib/sprockets/uri_tar.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        @scheme   = uri_array.shift
+        @scheme   << "://".freeze
+        @path     = uri_array.join("".freeze)
+      else
+        @scheme = "".freeze
+        @path   = uri
+      end
+    end
+
+
```
Comment (rafaelfranca) [inline lib/sprockets/uri_tar.rb]: :scissors:

File: lib/sprockets/loader.rb (hunk 5 above)
Comment (rafaelfranca) [inline lib/sprockets/loader.rb]: Ah, just saw above

**Comment by rafaelfranca** [comment]
LGTM, just a few minor comments

**Pull request merged**